vowpalwabbit/sampling.cc

```diff
@@ -0,0 +1,95 @@
+#include "hash.h"
+
+#include <string.h>
+#include <algorithm>
+#include <stdexcept>
+#include <numeric>
+
+#include "sampling.h"
+
+namespace exploration
+{
+  const uint64_t a = 0xeece66d5deece66dULL;
+  const uint64_t c = 2147483647;
+
+  const int bias = 127 << 23;
+
+  union int_float
+  {
+    int32_t i;
+    float f;
+  };
+
+  // uniform random between 0 and 1
+  float uniform_random_merand48(uint64_t initial)
+  {
+    initial = a * initial + c;
+    int_float temp;
+    temp.i = ((initial >> 25) & 0x7FFFFF) | bias;
+    return temp.f - 1;
+  }
+
+  sample sample_from_pdf(const char* seed, const float* pdf, uint32_t len)
+  {
+    uint64_t seed_hash = uniform_hash(seed, strlen(seed), 0);
+    return sample_from_pdf(seed_hash, pdf, len);
+  }
+
+  sample sample_from_pdf(uint64_t seed, const float* pdf, uint32_t len)
```

Comment:
If we want to be consistent, should this also have an STL-style iterator interface?

Reply:
I'd favro the STL-style interface

Reply:
done